riscv/runtime/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+    while i + 3 < n as isize {
+        *((dest.offset(i)) as *mut u32) = *((src.offset(i)) as *mut u32);
+        i += 4;
+    }
+    if i < n as isize {
+        let value = *((src.offset(i)) as *mut u32);
+        let dest_value = (dest.offset(i)) as *mut u32;
+        let mask = (1 << (((n as isize - i) * 8) as u32)) - 1;
+        *dest_value = (*dest_value & !mask) | (value & mask);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __runtime_memset(dest: *mut u8, val: u8, n: usize) -> *mut u8 {
+    let mut i: isize = 0;
+    let pattern = (val as u32) << 24 | (val as u32) << 16 | (val as u32) << 8 | (val as u32);
+
+    // Prologue: set bytes in the first non-aligned word
+    let first_word_offset = (dest as u32) % 4;
+    if first_word_offset != 0 {
+        let aligned_dest = dest.offset(-(first_word_offset as isize)) as *mut u32;
+        let mask = (1 << ((first_word_offset * 8) as u32)) - 1;
+        *aligned_dest = (*aligned_dest & mask) | (pattern & !mask);
+        i += 4 - first_word_offset as isize;
+    }
+
+    // Set all full words
+    while i + 3 < n as isize {
+        *((dest.offset(i)) as *mut u32) = pattern;
+        i += 4;
+    }
+
+    // last part: set the final word
+    if i < n as isize {
+        let dest_value = (dest.offset(i)) as *mut u32;
+        let mask = (1 << (((n as isize - i) * 8) as u32)) - 1;
+        *dest_value = (*dest_value & !mask) | (pattern & mask);
+    }
+
+    dest
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __runtime_memcmp(dest: *mut u8, src: *const u8, n: usize) -> bool {
+    if (src as u32) % 4 == (dest as u32) % 4 {
+        return memcmp_aligned(dest, src, n);
+    }
+
+    for i in 0..n {
+        if *dest.offset(i as isize) != *src.offset(i as isize) {
+            return false;
+        }
+    }
+    true
+}
+
+pub unsafe fn memcmp_aligned(dest: *mut u8, src: *const u8, n: usize) -> bool {
+    let mut i: isize = 0;
+
+    // Prologue: set bytes in the first non-aligned word
+    let first_word_offset = (dest as u32) % 4;
+    if first_word_offset != 0 {
+        let value = *((src.offset(-(first_word_offset as isize))) as *mut u32);
+        let aligned_dest = dest.offset(-(first_word_offset as isize)) as *mut u32;
+        let mask = (1 << ((first_word_offset * 8) as u32)) - 1;
+        if (*aligned_dest & !mask) != (value & !mask) {
+            return false;
+        }
+        i += 4 - first_word_offset as isize;
+    }
+
+    while i + 3 < n as isize {
+        if *((dest.offset(i)) as *mut u32) != *((src.offset(i)) as *mut u32) {
+            return false;
+        }
+        i += 4;
+    }
+
+    // last part: set the final word
+    if i < n as isize {
+        let value = *((src.offset(i)) as *mut u32);
+        let dest_value = (dest.offset(i)) as *mut u32;
+        let mask = (1 << (((n as isize - i) * 8) as u32)) - 1;
+        if (*dest_value & mask) != (value & mask) {
+            return false;
+        }
+    }
+    true
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Does this work? Otherwise, I think we can add some tests in the place where the keccak test is.

Reply:
no it doesn't work, and it shouldn't have been committed. This was an attempt to get it to work with `qemu-riscv32-static` but it won't compile because the `#[no_std]` flag is mandatory as `riscv32imc-unknown-none-elf` requires it, and rustup does not have `riscv32imc-unknown-linux-gnu`.